### PR TITLE
Power token v2

### PR DESCRIPTION
**Pull request by pete-ipor**

### CODE REVIEW

**Review by amydlarz — COMMENTED** [review]
V1

File: contracts/interfaces/IStakeService.sol [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    function stakeLpTokens(
+        address onBehalfOf,
+        address[] calldata lpTokens,
+        uint256[] calldata lpTokenAmounts
+    ) external;
+
+    function unstakeLpTokens(address[] calldata lpTokens, uint256[] calldata lpTokenAmounts)
+        external;
+
+    function stakeIporToken(address onBehalfOf, uint256 iporTokenAmount) external;
```
Comment (amydlarz) [inline contracts/interfaces/IStakeService.sol:14]: stakeIporToken  -> maybe stakeProcotolToken ?

File: contracts/interfaces/IFlowsService.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.17;
+
+interface IFlowsService {
```
Comment (amydlarz) [inline contracts/interfaces/IFlowsService.sol:4]: maybe DelegationRewardsService or DelegationStakingService?

File: contracts/mining/LiquidityMiningInternalV2.sol [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    ILiquidityMiningInternalV2
+{
+    using SafeCast for uint256;
+    using SafeCast for int256;
+
+    address public immutable ROUTER_ADDRESS; // Router address
+
+    // @deprecated do not use this
+    address internal _powerToken;
+    address internal _pauseManager;
```
Comment (amydlarz) [inline contracts/mining/LiquidityMiningInternalV2.sol:33]: what about a new approach with pausing?

File: contracts/mining/LiquidityMiningV2.sol [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+            }
+        }
+        return rewards;
+    }
+
+    function updateIndicators(address account, address[] calldata lpTokens)
+        external
+        override
+        nonReentrant
+        whenNotPaused
```
Comment (amydlarz) [inline contracts/mining/LiquidityMiningV2.sol:117]: it can be called by all or only router?

File: contracts/mining/LiquidityMiningV2.sol [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        override
+        returns (LiquidityMiningTypes.DelegatedPwTokenBalance[] memory balances)
+    {
+        uint256 lpTokensLength = lpTokens.length;
+        balances = new LiquidityMiningTypes.DelegatedPwTokenBalance[](lpTokensLength);
+        address lpToken;
+
+        for (uint256 i; i != lpTokensLength; ++i) {
+            lpToken = lpTokens[i];
+            require(_lpTokens[lpToken], Errors.LP_TOKEN_NOT_SUPPORTED);
```
Comment (amydlarz) [inline contracts/mining/LiquidityMiningV2.sol:45]: if we use validate lptokens then for account arg should be also validated?

File: contracts/services/StakeService.sol [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        ILiquidityMiningV2(LIQUIDITY_MINING_ADDRESS).addLpTokens(updateLpTokens);
+    }
+
+    function unstakeLpTokens(address[] calldata lpTokens, uint256[] calldata lpTokenAmounts)
```
Comment (amydlarz) [inline contracts/services/StakeService.sol:69]: onBehalfOf?

File: test/PowerTokensSystem.sol [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+import "./TestCommons.sol";
+import "../contracts/mining/LiquidityMiningV2.sol";
+import "../contracts/tokens/PowerTokenV2.sol";
+import "../contracts/lens/LiquidityMiningLens.sol";
+import "../contracts/services/StakeService.sol";
+import "../contracts/services/FlowsService.sol";
+import "../contracts/router/PowerTokenRouter.sol";
+import "../contracts/lens/PowerTokenLens.sol";
+
+contract PowerTokensSystem is TestCommons {
```
[the remark this answers is not in the document]
Comment (pete-ipor) [reply test/PowerTokensSystem.sol:18]: `PowerTokensTestsSystem`

File: contracts/services/StakeService.sol [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+import "../libraries/errors/Errors.sol";
+import "../interfaces/ILiquidityMiningV2.sol";
+import "../interfaces/IPowerTokenV2.sol";
+
+contract StakeService is IStakeService {
+    using SafeERC20 for IERC20;
+
+    address public immutable LIQUIDITY_MINING_ADDRESS;
+    address public immutable POWER_TOKEN_ADDRESS;
+    address public immutable STAKED_TOKEN_ADDRESS;
```
Comment (amydlarz) [inline contracts/services/StakeService.sol:16]: STAKED_TOKEN_ADDRESS -> PROTOCOL_TOKEN_ADDRESS? why here is a prefix STAKED?

File: contracts/lens/LiquidityMiningLens.sol [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+import "../interfaces/types/LiquidityMiningTypes.sol";
+import "../interfaces/ILiquidityMiningLens.sol";
+import "../interfaces/ILiquidityMiningV2.sol";
+
+contract LiquidityMiningLens is ILiquidityMiningLens {
+    address public immutable LIQUIDITY_MINING;
+
+    constructor(address liquidityMining) {
+        LIQUIDITY_MINING = liquidityMining;
```
Comment (amydlarz) [inline contracts/lens/LiquidityMiningLens.sol:12]: require(liquidityMining != address(0), Errors.WRONG_ADDRESS);

File: contracts/router/PowerTokenRouter.sol [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        address liquidityMiningAddress;
+        address powerTokenAddress;
+        address liquidityMiningLens;
+        address stakeService;
+        address miningService;
+        address powerTokenLens;
+    }
+
+    constructor(DeployedContracts memory deployedContracts) {
+        LIQUIDITY_MINING_LENS = deployedContracts.liquidityMiningLens;
```
Comment (amydlarz) [inline contracts/router/PowerTokenRouter.sol:33]: validation that addresses are not zero?

File: contracts/router/AccessControl.sol [hunk 11]
```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.17;
+
+import "../libraries/errors/Errors.sol";
+
+contract AccessControl {
```
Comment (amydlarz) [inline contracts/router/AccessControl.sol:6]: abstract

File: contracts/router/PowerTokenRouter.sol [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            // block because it will not return to Solidity code. We overwrite the
+            // Solidity scratch pad at memory position 0.
+            calldatacopy(0, 0, calldatasize())
+
+            // Call the implementation.
+            // out and outsize are 0 because we don't know the size yet.
+            result := delegatecall(gas(), implementation, 0, calldatasize(), 0, 0)
+
+            // Copy the returned data.
+            returndatacopy(0, 0, returndatasize())
```
[the remark this answers is not in the document]
Comment (pete-ipor) [reply contracts/router/PowerTokenRouter.sol:123]: yes

File: contracts/router/PowerTokenRouter.sol [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+            }
+        }
+    }
+
+    function batchExecutor(bytes[] calldata calls) external {
+        uint256 length = calls.length;
+        for (uint256 i; i != length; ) {
+            bytes4 sig = bytes4(calls[i][:4]);
+            address implementation = getRouterImplementation(sig);
+            implementation.functionDelegateCall(calls[i]);
```
[the remark this answers is not in the document]
Comment (pete-ipor) [reply contracts/router/PowerTokenRouter.sol:146]: for simplicity

**Pull request merged**